crates/jxl-oxide/tests/decode.rs

```diff
@@ -0,0 +1,115 @@
+                let mut marker = 0u8;
+                expected.read_exact(std::slice::from_mut(&mut marker)).unwrap();
+                if marker != 0 {
+                    panic!();
+                }
+
+                let color_channels = frame.color_channels();
+                let extra_channels = frame.extra_channels();
+                assert_eq!(fixture_header.channels as usize, color_channels.len() + extra_channels.len());
+
+                for grid in color_channels.iter().chain(extra_channels.iter().map(|ec| ec.grid())) {
+                    assert_eq!(fixture_header.width as usize, grid.width());
+                    assert_eq!(fixture_header.height as usize, grid.height());
+                    let buf = grid.buf();
+                    let mut expected_buf = vec![0u8; grid.width() * grid.height() * 2];
+                    expected.read_exact(&mut expected_buf).unwrap();
+
+                    for (&actual_float, expected) in buf.iter().zip(expected_buf.chunks_exact(2)) {
+                        let expected = u16::from_le_bytes([expected[0], expected[1]]);
+                        let actual = (actual_float * 65535.0 + 0.5) as u16;
+                        assert_eq!(expected, actual);
+                    }
+                }
+            },
+            jxl_oxide::RenderResult::NeedMoreData => panic!(),
+            jxl_oxide::RenderResult::NoMoreFrames => {
+                let mut marker = 0u8;
+                expected.read_exact(std::slice::from_mut(&mut marker)).unwrap();
+                if marker != 0xff {
+                    panic!();
+                }
+                break;
+            },
+        }
+    }
+}
+
+macro_rules! test {
+    ($($(#[$attr:meta])* $name:ident),* $(,)?) => {
+        $(
+            #[test]
+            $(#[$attr])*
+            fn $name() {
+                let mut path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+                path.push("tests/decode");
+                path.push(stringify!($name));
+
+                let mut input_jxl = path.clone();
+                input_jxl.push("input.jxl");
+                let input = std::fs::read(input_jxl).expect("Failed to read file");
+
+                let mut output_buf = path.clone();
+                output_buf.push("output.buf.zst");
+
+                match std::fs::File::open(output_buf) {
+                    Ok(expected) => {
+                        let expected = zstd::Decoder::new(expected).expect("Failed to open Zstandard stream");
+                        decode(&input, expected);
+                    },
+                    Err(_) => {
+                        let mut output_url = path.clone();
+                        output_url.push("output.buf.zst.url");
+                        let output_url = std::fs::read_to_string(output_url).expect("Failed to get fixture URL");
+
+                        let resp = reqwest::blocking::get(output_url.trim()).expect("Failed to connect");
```

Comment:
We would probably want to do caching here, like in `conformance.rs`

Reply:
Good point! Added caching for downloaded image buffers.